### ISSUE TITLE
How to run an OperationProcessor with complex constructor using console?

### ISSUE THREAD

**Issue by Nensec** [RSuter/NSwag#1290]
I've made a custom operationprocessor implementation, but it has a complex constructor. This is all fine in the WebAPI side because I manually instantiate it, but how to do the same for nswag.exe using a *.nswag configuration file?

**Comment by RSuter** [comment]
Can you create a new class inheriting the one with the complex ctor and initialize it with a default ctor and "default" parameters for the base?

**Comment by Nensec** [comment]
Sadly no, I should have probably mentioned that the constructor parameter is the WebApiToSwaggerGeneratorSettings class. Apologies.

**Comment by RSuter** [comment]
Ah, ok.
My plan is to implement something like a ISwaggerGeneratorSettingsFactory interface which generates the settings with all these processors etc. and then in the CLI settings you can set the factory type name. Also this factory can then be used for the UseSwaggerUi() methods. This way you would be much more flexible in the CLI and you can share the settings between the runtime generator and CLI generator....

What do you think?

**Comment by Nensec** [comment]
Sounds great in theory, I hope it doesn't tangle up existing code and flow too much.

I was hoping, perhaps in the meantime, to have a `SwaggerGeneratorSettings` property on the `OperationProcessorContext`, or on the `SwaggerGenerator`.

The `SwaggerGenerator` already takes the `JsonSchemaGeneratorSettings` as a constructor parameter, but is not a public property. If that would have been the case I could have tried casting it to the correct type.

**Comment by RSuter** [comment]
Yes, this would be implemented as a new feature - so should not be a breaking change for existing users...

Yes, we could expose settings in the context, but because it is not immutable, people may begin to manipulate it in a processor - which is very bad :-)

**Comment by RSuter** [comment]
So in your custom processor, you just need the settings (its the only ctor parameter)?

**Comment by Nensec** [comment]
Fair enough, maybe a read-only version of the settings class would be an option for that (but I realize that adds quite a bit of work too)

Yes, it is the only parameter.

**Comment by RSuter** [comment]
Ok, let's just add settings to the context. Seems to be a useful addition.

**Comment by Nensec** [comment]
Awesome. That should bridge the gap, at least for me, until the proposed factory implementation.

**Comment by RSuter** [comment]
Yes, the factory implementation will be much more effort - it involves a lot of places (generator, tooling, cli, etc.)

**Comment by Nensec** [comment]
I should be able to cast it and grab the needed things using that. I won't be able to test it out until next week however, did not quite expected it to be implemented this quickly heh

**Comment by RSuter** [comment]
Cast it to what?

**Comment by Nensec** [comment]
The property you made public is the base class of the `WebApiToSwaggerGeneratorSettings`, but the flow should provide an implementation of said class. So to get the settings my operationprocessor wants I'd need to cast it to that type first : )

**Comment by RSuter** [comment]
v11.17.2

**Comment by Nensec** [comment]
Works great, thanks!
